### ISSUE TITLE
Provide a set of common built-in transforms

### ISSUE THREAD

**Issue by doug-walker** [AcademySoftwareFoundation/OpenColorIO#980]
Currently all OCIO color spaces need to be provided using external LUT files (except for a few trivial possibilities that may be defined in the config, e.g. using an ExponentTransform). It would be useful to have some built-in transforms that OCIO could generate in memory and on demand. This would provide a number of advantages:

1)  It would be possible to define configs that don’t rely on external LUT files and hence would not break if these files are missing or unreadable from a given machine.
2)  Even where some LUTs are needed, not needing LUTs for the built-in transforms will make config packages smaller.
3)  It would make configs easier to create since authors would only have to create LUTs for transforms that are unique to their project and not have to worry about commonly used transforms.

Of course, there will be no requirement to use any of the built-in transforms. If the config author does not use them in the config then they won’t show up in application menus.  The proposal takes the approach of defining built-in transforms rather than built-in ColorSpaces to avoid name conflicts and to make it clear that it is up to the config author to "opt in" by using them in a ColorSpace definition.

Additional transform elements may be inserted by the config author if the reference space differs from what a given built-in transform expects.

The proposal is to implement all of the standard "building block" transforms used in modern color pipelines, so roughly this would be equivalent to the CTF collection that currently ships with Flame or Maya. This includes matrices for the common primary sets, linearization LUTs for all the main camera log spaces, interchange color spaces, video and digital cinema color spaces, etc.

Please note that this proposal does not add any new Ops, it simply provides a way to create transforms using the existing Ops.  So, just like with FileTransform, ultimately the BuiltinTransform results in the well-known existing Ops that are processed as usual.

Config Details

Here are some examples of how to refer to a built-in transform in a Config file. This would be used in from_reference or to_reference definitions, like FileTransform, ColorSpaceTransform, etc.:
```
   - !<BuiltInTransform> {name: CanonLog3_to_linear}
   - !<BuiltInTransform> {name: ST-2084_EOTF, direction: inverse}
   - !<BuiltInTransform> {name: Red_Log3G10_RedWideGamut_to_ACES2065-1}
```
The name argument would not be case-sensitive.  A function will be provided to iterate over all existing names.

**Comment by zachlewis** [comment]
I love this proposal. This would be so helpful for so many reasons... to the end that OCIO configs are human-readable and writeable, having an unambiguous, universal shorthand for commonly used transforms would do wonders for config reliability, validity, portability, and accessibility. I can't tell you how many times I've diagnosed configs with mismatched transfer functions and gamut transforms. It's a problem we've tried to mitigate with a library of very explicitly-named spimtx and spi1ds -- but, in practice, the fear and uncertainty surrounding color pipeline setup results in malformed colorspaces copied from show to show. 

We've been looking forward to using CLF / CTF as 'building blocks' in our pipelines, which would all but eliminate unintentionally combining one camera's CCTF with another's gamut transform. But having a built-in set of dependency-free transforms would do wonders for troubleshooting, inter-vendor communication, show setup... all sorts of things.

This would be especially useful for commercials....


Hey, do you think it would be possible to permit a FileTransform to failover to a given BuiltInTransform?

**Comment by doug-walker** [comment]
Thanks for the feedback @zachlewis!  Interesting point about a failover.  We will put together a PR for the BuiltinTransform.  It should be straight forward for someone to extend FileTransform to call that after exhausting its normal search path.

**Comment by zachlewis** [comment]
Thanks @doug-walker -- looking forward to the PR, whenever you guys get around to it! I'll open a separate thread afterwards specifically for the FileTransform failover business -- I'm not much for C++, but it does seem like a straight forward enough challenge, as you say, so I'll definitely try my hand.

You guys aren't planning a "NoOp" or "Null" BuiltInTransform by any chance, are you? If not, I'll make that part of the FileTransform thread. I can't really think of when a NoOp would be useful in other circumstances, but it would be a pretty slick way to have a FileTransform silently pass when nothing turns up in the search path, kinda like the way views can bypass a look with the "|" operator. (Our less-slick current practice involves pre-publishing nulls to the least granular part of our search path, which is overhead we could do without)

Thanks again! I really appreciate all your hard work...!

**Comment by doug-walker** [comment]
Zach, yes we will definitely have an "identity" Builtin that is just an identity matrix.

**Comment by sobotka** [comment]
Can we please supply XYZ first? Vastly more useful than going to ACES for a colour management tool.

**Comment by doug-walker** [comment]
Closing, as the basic functionality along with a set of example transforms (the ACES Color Space Conversion transforms) was implemented in PR #1006.  There is much more work to do with adding individual built-in transforms, but that is better tracked via separate issues.
